### ISSUE TITLE
Change system to use lessql

### ISSUE THREAD

**Issue by ddurieux** [glpi-project/glpi#942]
I have created a POC with use http://lessql.net/

You can test it: https://github.com/ddurieux/glpi/tree/master%2Blessql/
You can install (with databases MySQL, postfgresql and sqlite3) and create a computer

So the goal is to remove all joins and made simple queries, seems works nicely.

That we will do is to modify all queries in code to use an abstraction in GLPI and this abstraction layer use lessql to made queries.

**Comment by orthagh** [comment]
After  a (little) review and some discussions with @wawax and @trasher, i agree for adding this enhancement for the next major release (9.2 ?).

- Ok for lessql (support for postgres/mysql/sqlite, do we need another db support ?)
- Ok for jquery query builder, there is some losses (from current search form) we may need to resolve (first NOT operator, meta ?)
- I need more opinions on phinx, if i understand well, you push this lib for rollback support in migrations.
@trasher mentions rollbacks on mysql could not be exhaustive.

For the continuation of this task, we need: 
- a shared and full spec for this change, we can do it in a pr on the [docdec repo](https://github.com/glpi-project/docdev) (write for other dev enabled) if you agree.
- identify possible losses with the new system and discuss about them.
- get a scenario to migrate existing mysql queries in the code (apart Search class)
- linked to previous point, enhance db.function and dbmysql.class (need to be renamed) for adding common queries.
- enhance current poc to support mysql db
- write documentation (for plugins developers) on how to use the new system.

**Comment by ddurieux** [comment]
For rollback in phinx, it's a feature, the great feature is to manage migration for mysql, postgres easily.

enhance current poc to support mysql db -> the POC works on postgres AND mysql

Ok, I will begin to write specs.

Thanks

**Comment by trasher** [comment]
Indeed, I do mention some limitations about database transaction rollbacks on mysql, due to autocommit:
https://dev.mysql.com/doc/refman/5.7/en/implicit-commit.html

Well, for all other statements, and for other engines; this will work as expected. Users should continue to backup before migrate, but they already do that, didn't they ;)

I do remember I've already take a look at phinx; which I decide not to use, but I can't remember why exactly. I've take a look at their changelog; and maybe it was before postgresql support has been implemented; I've looked at it years ago!

**Comment by orthagh** [comment]
I did further test this evening and i have some concerns about this changes in fact.

For lessql, i have issues:

1- the lack of join support. This seems a big one.
I have a lot of doubts about their finding and traversal method and if it fits well to glpi.
We could have a lot a associated tables currently in some queries, and the possibility to let the users add a lot of fields in their query could end to a very bad performance with their system.
For 4 tables traversal like their example, i'm pretty sure that it could be fast, in our case we have more than dozen of tables involved in search queries (think about network ports for example). i not convinced by their "WHERE field IN(...)".
Actually our search engine is pretty fast when correctly used (don't sort on text fields when not necessary, prefer sort on id and never do text search when it's possible to use IS operator) and very flexible. I think we'll also have big losses in flexibility with this lib.

2- Also i can't find a way to support existing meta-criteria.
We can't drop them ! A lot of our users have usage of this feature. 
The best use case i can't find is the deploy feature of fusioninventory.
For example, some of our customers (in Teclib) use a simple criteria collection for computers finding in dynamic groups (simplified for the example) to deploy softwares:
- state IS "in production" 
- AND software - version IS NOT Firefox ESR 52

3- The rewrite needed for all queries of core and plugins.
We'll introduce a major change with this. It will be a pain to get a full up-to-date plugins list and no regression in the core.

The main change needed (and the reason for my acceptation) is the support of postgreSQL (and maybe SQlite). An alternate possibility (and a suggestion) would be to add a pdo abstraction to glpi db management.
I think the work to make queries compatibles with all engines is less by far than rewrite in lessql.
Here is a good article on the main differences to adapt our queries: https://evertpot.com/writing-sql-for-postgres-mysql-sqlite/



About jquery Querybuilder, i have also issues but i'm more confident about it:
My main concern is still the meta criteria. We have two "plus" buttons actually
Also, the lib lacks negative operator (in theirs blue buttons).
I read their documentation about plugins support, i think i could develop into a separate plugin our needs.
I like a lot this component, their advanced features (sorting rules, unique filters) are very good and we could add theses in a advanced mode.
Necessarily, the grouping or rules is a must have feature.



About phinx, i didn't do any tests this evening, will do next week.


Sorry for this late answer (and my shifting opinion). I passed my afternoon (and a big part of my evening) to test all of theses and i can't reassure me about this change.
I thank you for the work (on poc and spec), it not a loss, the discussion about this change start and some of your concept need to be keeped.

**Comment by ddurieux** [comment]
I have mades tests too with glpi_state (5 states) and glpi_computer (100 000 computers), benchmark 200 times each function with limit of 100 first computers:

* pdo_simple          66 ms
* pdo_jointure         67 ms
* mysqli_jointure     75 ms
* lessql_simple       2210 ms

So yes, seems lessql simple to use but not have good performances

**Comment by remicollet** [comment]
Performance and memory usage is a huge issue.

IMHO, some work are mandatory and can be done to prepare other change (aka, split this proposal)
* use NULL for dropdown
* use BOOLEAN instead of tinyint for bool (don't know if really needed)
* ensure whe use integer and no string for numerical value
* etc

**Comment by tomolimo** [comment]
Regarding performance: the simple switch to innodb gives a notable gain in response time.

**Comment by ddurieux** [comment]
The tests are done with innodb

**Comment by tomolimo** [comment]
Ok

**Comment by ddurieux** [comment]
Begin write specs:
* https://github.com/glpi-project/docdev/tree/sphinx+specs
* http://glpi-developer-documentation.readthedocs.io/en/sphinx-specs/specifications/index.html

**Comment by cedric-anne** [comment]
Hi,

Looks like LessQL is not a really active project. I am not sure that using it is a good idea.

Regards
